src/scripts/cfo.zig

```diff
@@ -0,0 +1,484 @@
+//! Continuous Fuzzing Orchestrator.
+//!
+//! We have a number of machines which run
+//!
+//!     git clone https://github.com/tigerbeetle/tigerbeetle && cd tigerbeetle
+//!     while True:
+//!         git fetch origin && git reset --hard origin/main
+//!         ./scripts/install_zig.sh
+//!         ./zig/zig build scripts -- cfo
+//!
+//! By modifying this script, we can make those machines do interesting things.
+//!
+//! The primary use-case is fuzzing: `cfo` runs a random fuzzer, and, if it finds a failure, it is
+//! recorded in devhubdb.
+//!
+//! Specifically:
+//!
+//! CFO keeps `args.concurrency` fuzzes running at the same time. For simplicity, it polls currently
+//! running fuzzers for completion every second in a fuzzing loop. A fuzzer fails if it returns
+//! non-zero error code.
+//!
+//! The fuzzing loops runs for `args.budget_minutes`. To detect hangs, if any fuzzer is still
+//! running after additional `args.hang_minutes`, it is killed (thus returning non-zero status and
+//! recording a failure).
+//!
+//! It is important that the caller (systemd typically) arranges for CFO to be a process group
+//! leader. It is not possible to reliably wait for (grand) children with POSIX, so its on the
+//! call-site to cleanup any run-away subprocesses
+//!
+//! After the fuzzing loop, CFO collects a list of seeds, some of which are failing. Next, it
+//! merges, this list into previous set of seeds (persisting seeds is to be implemented, at the
+//! moment the old list is always empty).
+//!
+//! Rules for merging:
+//!
+//! - Keep seeds for at most `commit_count_max` distinct commits.
+//! - Prefer fresher commits (based on commit time stamp).
+//! - For each commit and fuzzer combination, keep at most `seed_count_max` seeds.
+//! - Prefer failing seeds to successful seeds.
+//! - Prefer older seeds.
+//!
+//! These rules ensure that in the steady state (assuming fuzzer's clock doesn't fail) the set of
+//! seeds is stable. If the clock goes backwards, there might be churn in the set of a seed, but the
+//! number of failing seeds will never decrease.
+
+const std = @import("std");
+const builtin = @import("builtin");
+const log = std.log;
+const assert = std.debug.assert;
+
+const stdx = @import("../stdx.zig");
+const flags = @import("../flags.zig");
+const fatal = flags.fatal;
+const Shell = @import("../shell.zig");
+
+pub const CliArgs = struct {
+    budget_minutes: u64 = 10,
+    hang_minutes: u64 = 30,
+    concurrency: u32 = 16,
```

Comment:
Might be nicer to default this to num_cpu_cores in Zig than in the calling script?